api/router/core/ephys.py

```diff
@@ -0,0 +1,158 @@
+
+import h5py
+import matplotlib.pyplot as plt
+import numpy as np
+from fastapi import APIRouter, Depends, Query, Response
+from fastapi.security import HTTPBearer
+from loguru import logger as L
+from numpy.typing import NDArray
+
+from api.core.api import ApiError, ApiErrorCode
+from api.exceptions import ContentEmpty
+from api.http.entity_core import (
+    EntityType,
+    RequestContext,
+    get_entitycore_client,
+    get_request_context,
+)
+from api.models.enums import MetaType
+from api.tools.plot_ephys import plot_nwb_ephys
+from api.utils.common import get_buffer
+from api.utils.trace_img import (
+    get_conversion,
+    get_rate,
+    get_unit,
+    select_element,
+    select_protocol,
+    select_response,
+)
+
+
+@dataclass
+class EphysData:
+    """Container for ephys data and its metadata."""
+
+    data: NDArray
+    unit: str
+    rate: float
+
+
+router = APIRouter(
+    prefix="/single-cell-experimental-trace",
+)
+require_bearer = HTTPBearer()
+
+
+async def get_ephys_content(
+    entity_id: uuid.UUID,
+    asset_id: uuid.UUID,
+    context: RequestContext = Depends(get_request_context),
+) -> str:
+    """Get the ephys file content from the entity core service."""
+    async with get_entitycore_client() as core_client:
+        download_url = await core_client.download_entity_asset(
+            entity_type=EntityType.SINGLE_CELL_EXPERIMENTAL_TRACE,
+            entity_id=entity_id,
+            asset_id=asset_id,
+            context=context,
+        )
+        return await core_client.get_asset_content(download_url)
+
+
+def extract_ephys_data(ephys_file: str) -> EphysData:
+    """Extract data from the ephys HDF5 file."""
+    with h5py.File(io.BytesIO(ephys_file), "r") as h5_handle:
+        h5_handle = h5_handle["data_organization"]
+        h5_handle = h5_handle[select_element(list(h5_handle.keys()), n=0)]
+        h5_handle = h5_handle[select_protocol(list(h5_handle.keys()))]
+        h5_handle = h5_handle[select_element(list(h5_handle.keys()), n=0, meta=MetaType.REPETITION)]
+        h5_handle = h5_handle[select_element(list(h5_handle.keys()), n=-3, meta=MetaType.SWEEP)]
+        h5_handle = h5_handle[select_response(list(h5_handle.keys()))]
+
+        # Get relevant data, unit, rate, and conversion factor
+        unit = get_unit(h5_handle)
+        rate = get_rate(h5_handle)
+        conversion = get_conversion(h5_handle)
+
+        # Retrieve and process the data
+        data = np.array(h5_handle["data"][:]) * conversion
+
+        return EphysData(data=data, unit=unit, rate=rate)
+
+
+def generate_plot(ephys_data: EphysData, dpi: Optional[int]) -> bytes:
+    """Generate a plot from the ephys data."""
+    try:
+        fig = plot_nwb_ephys(ephys_data.data, ephys_data.unit, ephys_data.rate)
+        buffer = get_buffer(fig, dpi)
+        return buffer.getvalue()
+    except Exception as ex:
+        raise ApiError(
+            message=f"Error while converting asset to buffer: {ex}",
```

Comment:
Maybe "Error while generating plot" to be explicit, otherwise the description is a bit vague.